### PR TITLE
Remove support for `vnode.state = ...`

### PR DESCRIPTION
**Pull request by isiahmeadows**
## Description

As previously discussed, this removes the `vnode._state` internal member added in #1746 to avoid breaking existing object components. Now that we're shooting for a semver-major update, let's trim this garbage! 😄 

## Motivation and Context

See [this comment in #1727](https://github.com/MithrilJS/mithril.js/issues/1727#issuecomment-289254750) and the PR description in #1746.

## How Has This Been Tested?

Ran the tests as usual. Nothing out of the blue popped up, so it looks like it should work. Note that the reassignment check isn't actually tested, but it was never documented in the first place, with the assumption in #1339 that it would be considered read-only.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have updated `docs/change-log.md`

### CODE REVIEW

**Comment by isiahmeadows** [comment]
And some shiny perf numbers, in case you're concerned about the new sanity checks:

Chrome:

```
Before:
test-perf.js:61 rerender identical vnode x 4,088,255 ops/sec ±1.01% (53 runs sampled)
test-perf.js:61 rerender same tree x 88,768 ops/sec ±3.22% (44 runs sampled)
test-perf.js:61 construct large VDOM tree x 27,674 ops/sec ±2.67% (55 runs sampled)
test-perf.js:61 mutate styles/properties x 15,085 ops/sec ±2.88% (50 runs sampled)
test-perf.js:61 repeated trees (recycling) x 228 ops/sec ±3.77% (40 runs sampled)
test-perf.js:61 repeated trees (no recycling) x 199 ops/sec ±8.05% (43 runs sampled)
test-perf.js:67 Completed perf tests in 91215ms

After:
test-perf.js:61 rerender identical vnode x 4,083,779 ops/sec ±0.53% (55 runs sampled)
test-perf.js:61 rerender same tree x 89,641 ops/sec ±1.07% (48 runs sampled)
test-perf.js:61 construct large VDOM tree x 27,678 ops/sec ±1.09% (56 runs sampled)
test-perf.js:61 mutate styles/properties x 15,720 ops/sec ±1.57% (49 runs sampled)
test-perf.js:61 repeated trees (recycling) x 230 ops/sec ±4.20% (39 runs sampled)
test-perf.js:61 repeated trees (no recycling) x 204 ops/sec ±8.44% (43 runs sampled)
test-perf.js:67 Completed perf tests in 90960ms
```

Safari:

```
Before:
[Log] rerender identical vnode x 1,130,259 ops/sec ±8.20% (31 runs sampled) (test-perf.js, line 61)
[Log] rerender same tree x 16,245 ops/sec ±4.18% (40 runs sampled) (test-perf.js, line 61)
[Log] construct large VDOM tree x 2,476 ops/sec ±4.95% (36 runs sampled) (test-perf.js, line 61)
[Log] mutate styles/properties x 11,157 ops/sec ±1.31% (54 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (recycling) x 2,441 ops/sec ±1.04% (57 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (no recycling) x 2,480 ops/sec ±0.55% (59 runs sampled) (test-perf.js, line 61)
[Log] Completed perf tests in 36843ms (test-perf.js, line 67)

After:
[Log] rerender identical vnode x 1,075,703 ops/sec ±8.15% (29 runs sampled) (test-perf.js, line 61)
[Log] rerender same tree x 14,633 ops/sec ±4.08% (40 runs sampled) (test-perf.js, line 61)
[Log] construct large VDOM tree x 2,304 ops/sec ±5.75% (41 runs sampled) (test-perf.js, line 61)
[Log] mutate styles/properties x 11,157 ops/sec ±1.35% (56 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (recycling) x 2,162 ops/sec ±2.66% (55 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (no recycling) x 2,449 ops/sec ±1.71% (56 runs sampled) (test-perf.js, line 61)
[Log] Completed perf tests in 36895ms (test-perf.js, line 67)
```

Node:

```
Before:
rerender identical vnode x 6,313,480 ops/sec ±1.34% (89 runs sampled)
rerender same tree x 99,987 ops/sec ±1.69% (80 runs sampled)
construct large VDOM tree x 24,881 ops/sec ±1.09% (84 runs sampled)
mutate styles/properties x 46,808 ops/sec ±1.88% (82 runs sampled)
repeated trees (recycling) x 3,580 ops/sec ±2.13% (76 runs sampled)
repeated trees (no recycling) x 3,631 ops/sec ±2.07% (77 runs sampled)
Completed perf tests in 37239ms

After:
rerender identical vnode x 6,324,495 ops/sec ±1.46% (89 runs sampled)
rerender same tree x 101,641 ops/sec ±1.14% (80 runs sampled)
construct large VDOM tree x 25,018 ops/sec ±0.97% (87 runs sampled)
mutate styles/properties x 47,173 ops/sec ±1.64% (80 runs sampled)
repeated trees (recycling) x 3,646 ops/sec ±2.06% (74 runs sampled)
repeated trees (no recycling) x 3,663 ops/sec ±2.18% (75 runs sampled)
Completed perf tests in 36867ms
```

File: docs/change-log.md [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 - API: `m.redraw()` is always asynchronous ([#1592](https://github.com/MithrilJS/mithril.js/pull/1592))
 - API: `m.mount()` will only render its own root when called, it will not trigger a `redraw()` ([#1592](https://github.com/MithrilJS/mithril.js/pull/1592))
+- API: `vnode.state = ...` is no longer supported. Instead, an error is thrown if `vnode.state` changes upon the invocation of a lifecycle hook.
```
Comment (tivac) [inline docs/change-log.md]: I don't think this is clear enough. Something like 

> Assigning a new object to `vnode.state` (`vnode.state = ...`) is no longer supported

would help me understand.

File: docs/vnodes.md [hunk 2]
```diff
@@ -74,7 +74,6 @@ Property   | Type                             | Description
 `dom`      | `Element?`                       | Points to the element that corresponds to the vnode. This property is `undefined` in the `oninit` lifecycle method. In fragments and trusted HTML vnodes, `dom` points to the first element in the range.
 `domSize`  | `Number?`                        | This is only set in fragment and trusted HTML vnodes, and it's `undefined` in all other vnode types. It defines the number of DOM elements that the vnode represents (starting from the element referenced by the `dom` property).
 `state`    | `Object?`                        | An object that is persisted between redraws. It is provided by the core engine when needed. In POJO component vnodes, the `state` inherits prototypically from the component object/class. In class component vnodes it is an instance of the class. In closure components it is the object returned by the closure.
-`_state`   | `Object?`                        | For components, a reference to the original `vnode.state` object, used to lookup the `view` and hooks. This property is only used internally by Mithril, do not use or modify it.
```
Comment (tivac) [inline docs/vnodes.md]: Good riddance.

**Comment by pygy** [comment]
I remember we discussed generating the state of non-components vnodes lazily using a getter (and thus keeping `_state` as a private backing store)... A corresponding setter could throw. I'll give it a try and look at perf numbers.

**Comment by isiahmeadows** [comment]
@pygy Here's the correct perf numbers for my current patch, run locally (I just copied my previous "Before" runs):

And some shiny perf numbers, in case you're concerned about the new sanity checks:

Chrome:

```
Before:
test-perf.js:61 rerender identical vnode x 4,088,255 ops/sec ±1.01% (53 runs sampled)
test-perf.js:61 rerender same tree x 88,768 ops/sec ±3.22% (44 runs sampled)
test-perf.js:61 construct large VDOM tree x 27,674 ops/sec ±2.67% (55 runs sampled)
test-perf.js:61 mutate styles/properties x 15,085 ops/sec ±2.88% (50 runs sampled)
test-perf.js:61 repeated trees (recycling) x 228 ops/sec ±3.77% (40 runs sampled)
test-perf.js:61 repeated trees (no recycling) x 199 ops/sec ±8.05% (43 runs sampled)
test-perf.js:67 Completed perf tests in 91215ms

After:
test-perf.js:61 rerender identical vnode x 3,786,353 ops/sec ±1.37% (52 runs sampled)
test-perf.js:61 rerender same tree x 88,690 ops/sec ±4.15% (51 runs sampled)
test-perf.js:61 construct large VDOM tree x 27,623 ops/sec ±2.81% (49 runs sampled)
test-perf.js:61 mutate styles/properties x 14,817 ops/sec ±3.16% (46 runs sampled)
test-perf.js:61 repeated trees (recycling) x 216 ops/sec ±7.26% (38 runs sampled)
test-perf.js:61 repeated trees (no recycling) x 202 ops/sec ±8.65% (45 runs sampled)
test-perf.js:67 Completed perf tests in 98331ms
```

Safari:

```
Before:
[Log] rerender identical vnode x 1,130,259 ops/sec ±8.20% (31 runs sampled) (test-perf.js, line 61)
[Log] rerender same tree x 16,245 ops/sec ±4.18% (40 runs sampled) (test-perf.js, line 61)
[Log] construct large VDOM tree x 2,476 ops/sec ±4.95% (36 runs sampled) (test-perf.js, line 61)
[Log] mutate styles/properties x 11,157 ops/sec ±1.31% (54 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (recycling) x 2,441 ops/sec ±1.04% (57 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (no recycling) x 2,480 ops/sec ±0.55% (59 runs sampled) (test-perf.js, line 61)
[Log] Completed perf tests in 36843ms (test-perf.js, line 67)

After:
[Log] rerender identical vnode x 1,025,451 ops/sec ±9.23% (29 runs sampled) (test-perf.js, line 61)
[Log] rerender same tree x 15,795 ops/sec ±5.91% (38 runs sampled) (test-perf.js, line 61)
[Log] construct large VDOM tree x 3,270 ops/sec ±11.42% (21 runs sampled) (test-perf.js, line 61)
[Log] mutate styles/properties x 11,745 ops/sec ±1.54% (57 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (recycling) x 2,232 ops/sec ±1.11% (59 runs sampled) (test-perf.js, line 61)
[Log] repeated trees (no recycling) x 2,194 ops/sec ±0.90% (60 runs sampled) (test-perf.js, line 61)
[Log] Completed perf tests in 37295ms (test-perf.js, line 67)
```

Node:

```
Before:
rerender identical vnode x 6,313,480 ops/sec ±1.34% (89 runs sampled)
rerender same tree x 99,987 ops/sec ±1.69% (80 runs sampled)
construct large VDOM tree x 24,881 ops/sec ±1.09% (84 runs sampled)
mutate styles/properties x 46,808 ops/sec ±1.88% (82 runs sampled)
repeated trees (recycling) x 3,580 ops/sec ±2.13% (76 runs sampled)
repeated trees (no recycling) x 3,631 ops/sec ±2.07% (77 runs sampled)
Completed perf tests in 37239ms

After:
rerender identical vnode x 6,236,248 ops/sec ±1.41% (90 runs sampled)
rerender same tree x 98,985 ops/sec ±1.08% (82 runs sampled)
construct large VDOM tree x 24,682 ops/sec ±0.82% (82 runs sampled)
mutate styles/properties x 46,168 ops/sec ±1.63% (82 runs sampled)
repeated trees (recycling) x 3,252 ops/sec ±4.20% (70 runs sampled)
repeated trees (no recycling) x 3,590 ops/sec ±1.88% (76 runs sampled)
Completed perf tests in 36648ms
```

**Comment by isiahmeadows** [comment]
@pygy I'll note that with yours, the perf hit *should* be much less if we decide to change it from an object literal to using a prototype. Note that this will be at the cost of code size, though.

I'd be okay with it if that were done, but I'm more of a +0 because of the almost inevitable code size jump.

File: docs/change-log.md (hunk 1 above)
Comment (isiahmeadows) [reply docs/change-log.md]: Fixed in a force push. 👍

**Comment by pygy** [comment]
Good point regarding the prototype... Changing to getters would however require a massive test update, since there are many tests that use POJO as vnodes...

**Comment by isiahmeadows** [comment]
And as much as that sounds like a good idea, [I strongly doubt @lhorie would get behind the necessary test changes for it](https://github.com/MithrilJS/mithril.js/pull/1198#issuecomment-237245567)... 😦

**Comment by isiahmeadows** [comment]
@pygy Have you run perf numbers yet? (Oh, and BTW, can this be merged now?)

**Pull request merged**